### PR TITLE
Refactor dorel

### PR DESCRIPTION
**Pull request by RonaldGalea**
model able to overfit validation set

### CODE REVIEW

File: architectures/models/SSDNet.py [hunk 1]
```diff
@@ -15,16 +15,16 @@ def __init__(self, in_channels, n_classes, k):
         self.k = k
         '''
         idea: adding a DWSC is cheap anyway, so might as well try to use it to 'specialize' features onto this particular output from the main stream
-        '''
+        # '''
```
Comment (pasandrei) [inline architectures/models/SSDNet.py]: Why are these quotation marks commented?

File: architectures/models/SSDNet.py [hunk 2]
```diff
@@ -53,11 +53,11 @@ def __init__(self, n_classes=3, k=6, width_mult=1):
         self.inv3 = InvertedResidual(inp=512, oup=256, stride=2, expand_ratio=0.25)
         self.out3 = OutConv(256, n_classes, k)
 
-        # fourth grid 2x2
+        # # fourth grid 2x2
```
Comment (pasandrei) [inline architectures/models/SSDNet.py]: Remove the double comment here and on lines 59 and 60

File: misc/print_stats.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    avg_grads, max_grads = torch.FloatTensor(avg_grads), torch.FloatTensor(max_grads)
+    avg_weigths, max_weigths = torch.FloatTensor(avg_weigths), torch.FloatTensor(max_weigths)
+
+    return torch.mean(avg_grads), torch.mean(max_grads), torch.mean(avg_weigths), torch.mean(max_weigths)
+
+# print stats
+
+
+def plot_grad_flow(model):
```
Comment (pasandrei) [inline misc/print_stats.py:46]: Is this new?

File: my_tests/test_tensor_unfolding.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import torch
```
Comment (pasandrei) [inline my_tests/test_tensor_unfolding.py]: I don't think that this is necessary to test the model

To be removed

File: run_tests.py [hunk 5]
```diff
@@ -1,8 +1,13 @@
-from my_tests import anchor_mapping, speed_test
+from my_tests import speed_test, test_nms, test_tensor_unfolding
 
+print('brtgrg')
```
Comment (pasandrei) [inline run_tests.py]: To be removed

File: run_tests.py [hunk 6]
```diff
@@ -1,8 +1,13 @@
...
-    anchor_mapping.test_anchor_mapping()
-
-def run_speed_test():
-    speed_test.measure_mobilenet()
+def run():
+    # def run_speed_test():
+    #     speed_test.measure_mobilenet()
+    #
+    # def run_nms_test():
+    #     print('afsf')
```
Comment (pasandrei) [inline run_tests.py]: Also to be removed

File: train/helpers.py [hunk 7]
```diff
@@ -111,87 +112,43 @@ def create_anchors():
...
+    #anchor_cnr = hw2corners(anchors[:, :2], anchors[:, 2:])
 
-# helper for dataset
+    return anchors, grid_sizes
 
 
-def prepare_gt(x, y):
+def prepare_gt(input_img, gt_target):
     '''
+    NUMA SEFU STIE CE SENTAMPLA AICI. NICI MACAR DUMNEZEU NU ARE IDEEE
```
Comment (pasandrei) [inline train/helpers.py:122]: MARE PLUS

nici io nu mai stiu ce se intampla aici

Comment (pasandrei) [reply train/helpers.py:122]: defapt, miam amintit ce face

File: train/loss_fn.py [hunk 8]
```diff
@@ -69,11 +78,20 @@ def ssd_loss(pred, targ, anchors, grid_sizes, device, params):
 
     # computes the loss for each image in the batch
     for idx in range(pred[0].shape[0]):
+
         pred_bbox, pred_class = pred[0][idx], pred[1][idx]
         gt_bbox, gt_class = targ[0][idx].to(device), targ[1][idx].to(device)
 
+        # assert that all tensors passed to ssd_1_loss are on GPU !!!!!!!!
+        # assert pred_bbox.is_cuda is True
+        # assert pred_class.is_cuda is True
```
Comment (pasandrei) [inline train/loss_fn.py]: Do we need these comments?

File: train/validate.py [hunk 9]
```diff
@@ -11,10 +10,12 @@ def evaluate(model, optimizer, anchors, grid_sizes, train_loader, valid_loader,
     '''
 
     print('Average loss this epoch: Localization: {}; Classification: {}'.format(
-        losses[2] / len(train_loader), losses[3] / len(train_loader)))
+        losses[2] / len(train_loader.dataset), losses[3] / len(train_loader.dataset)))
     print('Validation start...')
 
     l_val_loss, c_val_loss = 0, 0
+
+    BATCHES_TO_TEST = 4
```
Comment (pasandrei) [inline train/validate.py]: Why is this hardcoded?

**Review by pasandrei — COMMENTED** [review]
I added a few comments

File: architectures/models/SSDNet.py (hunk 1 above)
Comment (skiry) [reply architectures/models/SSDNet.py]: > Why are these quotation marks commented?

The quotes are not commented, the comment is quoted.

File: architectures/models/SSDNet.py [hunk 10]
```diff
@@ -97,5 +97,7 @@ def forward(self, x):
         x = self.inv5(x)
         _1bbox, _1class = self.out5(x)
 
-        return [torch.cat([_20bbox, _10bbox, _5bbox, _3bbox, _2bbox, _1bbox], dim=1),
-                torch.cat([_20class, _10class, _5class, _3class, _2class, _1class], dim=1)]
+        # bboxes prediction:  B x (20*20) * 6 x 4 ...
+        # class prediction:  B x (20*20) * 6 x 3 ...
+        return [torch.cat([_10bbox, _5bbox, _3bbox, _2bbox, _1bbox], dim=1),
```
Comment (skiry) [inline architectures/models/SSDNet.py:102]: So we prune the searching space by not considering smaller objects than ~10x10?

File: misc/postprocessing.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    if len(bounding_boxes.shape) == 1:
+        cv2.rectangle(image, (bounding_boxes[1], bounding_boxes[0]),
+                      (bounding_boxes[3], bounding_boxes[2]), color, 2)
+    else:
+        for (startX, startY, endX, endY) in bounding_boxes:
+            cv2.rectangle(image, (startY, startX), (endY, endX), color, 2)
+
+    # display the image
+    cv2.imshow(message, image)
+    cv2.waitKey(0)
```
Comment (skiry) [inline misc/postprocessing.py:94]: This algorithm is so well written, very nice!

**Review by skiry — APPROVED** [review]

File: architectures/models/SSDNet.py (hunk 10 above)
Comment (RonaldGalea) [reply architectures/models/SSDNet.py:102]: Vehicles and people would rarely be a good fit for the 20x20 grid, so we ditch that as it basically only worsens class imbalance

**Review by pasandrei — APPROVED** [review]
